src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1239 @@
+{
+	PruningInstance *instance;
+	Node *continueAt;
+} PendingPruningInstance;
+
+
+/*
+ * Data necessary to perform a single PruneShards().
+ */
+typedef struct ClauseWalkerContext
+{
+	Var *partitionColumn;
+	char partitionMethod;
+
+	/* ORed list of pruning targets */
+	List *pruningInstances;
+
+	/*
+	 * Partially built PruningInstances, that need to be completed. XXX
+	 */
+	List *pendingInstances;
+
+	/* PruningInstance currently being built, all elegible constraints are added here */
+	PruningInstance *currentPruningInstance;
+
+	/*
+	 * Information about function calls we need to perform. Re-using the same
+	 * FunctionCallInfoData, instead of using FunctionCall2Coll, is often
+	 * cheaper.
+	 */
+	FunctionCallInfoData compareValueFunctionCall;
+	FunctionCallInfoData compareIntervalFunctionCall;
+} ClauseWalkerContext;
+
+static void PrunableExpressions(Node *originalNode, ClauseWalkerContext *context);
+static bool PrunableExpressionsWalker(Node *originalNode, ClauseWalkerContext *context);
+static List * ShardArrayToList(ShardInterval **shardArray, int length);
+static int PerformValueCompare(FunctionCallInfoData *compareFunctionCall, Datum a,
+							   Datum b);
+static int PerformCompare(FunctionCallInfoData *compareFunctionCall);
+
+static List * PruneOne(DistTableCacheEntry *cacheEntry, ClauseWalkerContext *context,
+					   PruningInstance *prune);
+static List * PruneWithBoundaries(DistTableCacheEntry *cacheEntry,
+								  ClauseWalkerContext *context,
+								  PruningInstance *prune);
+static List * ExhaustivePrune(DistTableCacheEntry *cacheEntry,
+							  ClauseWalkerContext *context,
+							  PruningInstance *prune);
+static int UpperShardBoundary(Datum partitionColumnValue,
+							  ShardInterval **shardIntervalCache,
+							  int shardCount, FunctionCallInfoData *compareFunction,
+							  bool includeMin);
+static int LowerShardBoundary(Datum partitionColumnValue,
+							  ShardInterval **shardIntervalCache,
+							  int shardCount, FunctionCallInfoData *compareFunction,
+							  bool includeMax);
+
+
+/*
+ * PruneShards returns all shards from a distributed table that cannot be
+ * proven to be eliminated by whereClauseList.
+ *
+ * For reference tables, the function simply returns the single shard that the
+ * table has.
+ */
+List *
+PruneShards(Oid relationId, Index rangeTableId, List *whereClauseList)
+{
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	char partitionMethod = cacheEntry->partitionMethod;
+	Var *partitionColumn = NULL;
+	ClauseWalkerContext context = { 0 };
+	ListCell *pruneCell;
+	List *prunedList = NIL;
+	bool foundRestriction = false;
+
+	/* always return empty result if WHERE clause is of the form: false (AND ..) */
+	if (ContainsFalseClause(whereClauseList))
+	{
+		return NIL;
+	}
+
+	/* short circuit for reference tables */
+	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		return ShardArrayToList(cacheEntry->sortedShardIntervalArray,
```

Comment:
Hm. You have a point with the deep copy - but I think it'd make more sense to do that one layer up.  Otherwise we'll have to do a lot more work when ORing two PruneOne results.